comparison/loadtest.go

```diff
@@ -216,12 +223,43 @@ func initLoadTest(t *terraform.Terraform, buildCfg BuildConfig, dumpFilename str
 		return fmt.Errorf("invalid db engine %s", dpConfig.TerraformDBSettings.InstanceEngine)
 	}
 
+	resetBucketCmds := []localCmd{}
+	if s3BucketURI != "" && tfOutput.HasS3Bucket() {
+		deleteBucketCmd := localCmd{
+			msg:   "Emptying S3 bucket",
+			value: []string{"--profile", t.Config().AWSProfile, "s3", "rm", "s3://" + tfOutput.S3Bucket.Id, "--recursive"},
+		}
+
+		prepopulateBucketCmd := localCmd{
+			msg:   "Pre-populating S3 bucket",
+			value: []string{"--profile", t.Config().AWSProfile, "s3", "cp", s3BucketURI, "s3://" + tfOutput.S3Bucket.Id, "--recursive"},
+		}
+
+		resetBucketCmds = []localCmd{deleteBucketCmd, prepopulateBucketCmd}
+	}
+
 	if dumpFilename == "" {
 		cmds = append(cmds, startCmd, createAdminCmd, initDataCmd)
 	} else {
 		cmds = append(cmds, loadDBDumpCmd, startCmd)
 	}
 
+	// Resetting the buckets can happen concurrently with the rest of the remote commands,
+	// but we need to cancel them on return in case we return early (as when we receive from cancelCh)
+	resetBucketCtx, resetBucketCancel := context.WithCancel(context.Background())
+	defer resetBucketCancel()
+	resetBucketErrCh := make(chan error)
```

Comment:
The channel is unbuffered, so the routine would block until the channel is read. With a small buffer the goroutine could terminate (and run the deferred function) before line 279 is reached.

Reply:
Done in https://github.com/mattermost/mattermost-load-test-ng/pull/524/commits/91464b37647d1147f85bb21b5e2b64dc4a525624. But this does not affect the deferred function, right?

Reply:
You are right. I moved the `defer` into the go routine in my mind. Should have had more coffee. The buffered channel allows the go routine to exit and release its memory, though.